real_time/surface_src/java_src/BuggyRos/src/main/java/com/roboclub/robobuggy/ros/internal/MessageServer.java

```diff
@@ -31,14 +31,13 @@
 
 	private static MessageServer _master;
 	private static Lock singleton_lock = new ReentrantLock();
-	public static final int MAX_MESSAGE_QUEUE = 20;
 
 	private Map<String, List<Subscriber>> outbox_mapping = new HashMap<String, List<Subscriber>>();
 	private ReadWriteLock outbox_lock = new ReentrantReadWriteLock(true);
 
 	// Contains all messages that threads want sent.
 	// TODO: confirm threadsafety.
-	private LinkedBlockingQueue<Map.Entry<String, Message>> inbox = new LinkedBlockingQueue<Map.Entry<String, Message>>(MAX_MESSAGE_QUEUE);
+	private LinkedBlockingQueue<Map.Entry<String, Message>> inbox = new LinkedBlockingQueue<Map.Entry<String, Message>>();
```

Comment:
Dropping messages is a reasonable way to deal with the potential problem of backlogged messages, but the current implementation of dropping messages silently from the global message queue with no error message to the publisher or subscriber is not acceptable for a reliable system. This means any single publisher can DDOS the entire system on accident, and no one will know about it. Logs would show that publishers succeeded, but some subscriber callbacks just didn't happen. 

If we want to implement message dropping, it should be done on the local_inbox side (in Subscriber.java). We would want to drop the oldest message first. The Subscriber constructor could specify how long it would like the message queue to be. This lets the caller decide how messages could be dropped (GUI updates can drop all but the latest, kalman filter may want to keep more but not too many). 

The subscriber callback should be informed that messages are being dropped. This could be accomplished by the Publisher adding a sequence number to the messages it sends, that is unique per message and monotonically increasing. The subscriber (if they care) can check the sequence numbers to learn that messages have been dropped, and how many. It is then up to the Subscriber to log an error if necessary (GUI won't care, but kalman filter should). 

Currently, the framework has no instrumentation, so we don't know where the bottlenecks are. If the above was implemented, we could use that and make the Subscribers publish error messages in case bad things are happening. Making the instrumentation not affect performance would also be important, so we can fix real problems, not problems induced by instrumentation.